aws/resource_aws_network_interface_sg_attachment.go

```diff
@@ -0,0 +1,183 @@
+	// Get a lock to prevent races on other SG attachments/detatchments on this
+	// interface ID. This lock is released when the function exits, regardless of
+	// success or failure.
+	//
+	// The lock here - in the create function - deliberately covers the
+	// post-creation read as well, which is normally not covered as Read is
+	// otherwise only performed on refresh. Locking on it here prevents
+	// inconsistencies that could be caused by other attachments that will be
+	// operating on the interface, ensuring that Create gets a full lay of the
+	// land before moving on.
+	mk := "network_interface_sg_attachment_" + d.Get("network_interface_id").(string)
+	awsMutexKV.Lock(mk)
+	defer awsMutexKV.Unlock(mk)
+
+	sgID := d.Get("security_group_id").(string)
+	interfaceID := d.Get("network_interface_id").(string)
+
+	conn := meta.(*AWSClient).ec2conn
+
+	// Fetch the network interface we will be working with.
+	iface, err := fetchNetworkInterface(conn, interfaceID)
+	if err != nil {
+		return err
+	}
+
+	// Add the security group to the network interface.
+	log.Printf("[DEBUG] Attaching security group %s to network interface ID %s", sgID, interfaceID)
+
+	if sgExistsInENI(sgID, iface) {
+		return fmt.Errorf("security group %s already attached to interface ID %s", sgID, *iface.NetworkInterfaceId)
+	}
+	var groupIDs []string
+	for _, v := range iface.Groups {
+		groupIDs = append(groupIDs, *v.GroupId)
+	}
+	groupIDs = append(groupIDs, sgID)
+	params := &ec2.ModifyNetworkInterfaceAttributeInput{
+		NetworkInterfaceId: iface.NetworkInterfaceId,
+		Groups:             aws.StringSlice(groupIDs),
+	}
+
+	_, err = conn.ModifyNetworkInterfaceAttribute(params)
+	if err != nil {
+		return err
+	}
+
+	log.Printf("[DEBUG] Successful attachment of security group %s to network interface ID %s", sgID, interfaceID)
+
+	return resourceAwsNetworkInterfaceSGAttachmentRead(d, meta)
+}
+
+func resourceAwsNetworkInterfaceSGAttachmentRead(d *schema.ResourceData, meta interface{}) error {
+	sgID := d.Get("security_group_id").(string)
+	interfaceID := d.Get("network_interface_id").(string)
+
+	log.Printf("[DEBUG] Checking association of security group %s to network interface ID %s", sgID, interfaceID)
+
+	conn := meta.(*AWSClient).ec2conn
+
+	iface, err := fetchNetworkInterface(conn, interfaceID)
+	if err != nil {
+		return err
+	}
+
+	if sgExistsInENI(sgID, iface) {
+		d.SetId(fmt.Sprintf("%s_%s", sgID, interfaceID))
+	} else {
+		// The assocation does not exist when it should, taint this resource.
+		log.Printf("[WARN] Security group %s not associated with network interface ID %s, tainting", sgID, interfaceID)
+		d.SetId("")
+	}
+	return nil
+}
+
+func resourceAwsNetworkInterfaceSGAttachmentDelete(d *schema.ResourceData, meta interface{}) error {
+	// Get a lock to prevent races on other SG attachments/detatchments on this
+	// interface ID. This lock is released when the function exits, regardless of
+	// success or failure.
```

Comment:
I think when the lock is released is obvious from the code below, so the second sentence seems a bit redundant.